README.md
# App Tiny ERP

[![Deploy](https://github.com/ecomplus/app-tiny-erp/workflows/Deploy/badge.svg)](https://github.com/ecomplus/app-tiny-erp/actions?workflow=Deploy) [![Publish](https://github.com/ecomplus/app-tiny-erp/workflows/Publish/badge.svg)](https://github.com/ecomplus/app-tiny-erp/actions?workflow=Publish) [![License GPL](https://img.shields.io/badge/License-GPL-orange.svg)](https://opensource.org/licenses/GPL)

E-Com Plus app to integrate Tiny ERP

[CHANGELOG](https://github.com/ecomplus/app-tiny-erp/blob/master/CHANGELOG.md)

# E-Com Plus Application Starter

:dog: Boilerplate for E-Com Plus apps with [Firebase](https://firebase.google.com/) Cloud Functions and GitHub Actions.

## Getting started

1. Start creating a [Firebase project](https://console.firebase.google.com/):
    - Analytics is not needed;
    - Set a nice project name (ID) and remember it;

2. Enter the project, go to _databases_ page (on menu) and _create database_:
    - Just bypass with default production mode and rules;

3. Firebase free plan doesn't support sending external HTTP requests, so you'll need to upgrade to _Blaze_ (on demand) plan;

4. Get your Firebase token from CLI:
```bash
npm install -g firebase-tools
firebase login:ci
```

5. [Use this template](https://github.com/ecomplus/application-starter/generate) to generate a new repository for your application;

6. Go to your repository _settings_ tab and set the following _secrets_:
    - `FIREBASE_PROJECT_ID`: The ID (name) of your Firebase project;
    - `FIREBASE_TOKEN`: The token generated with `firebase-tools`;
    - `SERVER_OPERATOR_TOKEN`: Random (at least 16 bytes) admin token generated from CLI or [here](https://randomkeygen.com/);
    - `MARKET_TOKEN` (optional): Your E-Com Plus Market [personal token](https://market.e-com.plus/partners/account) to publish the app automatically;

## Next steps

Almost ready, time to :coffee: and code!

Edit [`functions/ecom.config.js`](functions/ecom.config.js) to set correct `app_id`, `title` and optionally more fields on base app body.

If you're not yet familiarized with this boilerplate, **read with attention the comments and instructions at the configuration file**. You can also setup procedures from there to specify the web-hooks your app should receive.

### E-Com common routes

All endpoints inside [`/ecom/`](functions/routes/ecom) path **MUST BE PRIVATE** (trey are by default), accessible only for E-Com Plus official servers or with `X-Operator-Token` header (equal to `SERVER_OPERATOR_TOKEN` env).

Requests to E-Com routes must have `X-Store-ID` header or `store_id` parameter on URL query string.

#### [`auth-callback`](functions/routes/ecom/auth-callback.js)

Handle [authentication callback request](https://developers.e-com.plus/docs/api/#/store/authenticate-app/authenticate-app) and save tokens to database, also create Store API procedures if configured.

> You can edit it to send custom requests to external server notifying new store installation or setup.

#### [`get-auth`](functions/routes/ecom/get-auth.js)

Returns Store API authentication data based on received Store ID, for external usage.

Sample response:

```json
{
  "application_id": "a00000000000000000000012",
  "application_app_id": 9000,
  "store_id": 1011,
  "authentication_id": "120000000000000000000012",
  "access_token": "eyJhbGciOi.eyYzdWIi.ZEONFh7HgQ"
}
```

> You should consume it if you want to use this boilerplate as an _authentication backend_ only. By getting this data (HTTP GET) you'll be able to run authenticated requests to [Store API](https://developers.e-com.plus/docs/api/#/store/) from your own server, with your preferred language, framework...

#### [`refresh-tokens`](functions/routes/ecom/refresh-tokens.js)

Call update service to start refreshing old access tokens.

#### [`webhook`](functions/routes/ecom/webhook.js)

Receives E-Com Plus notification POSTs for configured procedures.

> You may want to edit it to properly treat trigger body and do your custom stuff for each notification type.

#### [`modules/*`](functions/routes/ecom/modules)

Endpoints for [E-Com Plus Mods API](https://developers.e-com.plus/modules-api/).

> You should edit the respective ones for enabled modules (on `ecom.config.js`).

### Additional Store API handlers

If the app works with `admin_settings` (configuration), use [`getAppData`](functions/lib/store-api/get-app-data.js) to get the application `data`/`hidden_data` configured by merchant from dashboard through Store API ([Application Object](https://developers.e-com.plus/docs/api/#/store/applications/)).

We recommend treating Store API clients (40x) and server (50x) errors with [`errorHandling`](functions/lib/store-api/error-handling.js) abstraction.

### Application SDK

You may want to use [`appSdk`](https://github.com/ecomplus/application-sdk) to make custom authenticated requests to [Store API](https://developers.e-com.plus/docs/api/#/store/) or deeper coding out of our examples.

### Custom integration scripts

- Add third party libraries to [`functions/package.json`](functions/package.json) dependencies;
- Add custom web app routes by creating new files to [`functions/routes`](functions/routes) folder;
- Add new methods/handlers at [`functions/lib`](functions/lib) folder;

### Optional setup for [Market](https://market.e-com.plus/)

You may want to make your app public and available to install within E-Com Plus dashboard, to do this:

- Edit [`ecomplus-market.json`](ecomplus-market.json) to set a short description and category between _shipping_, _sales_, _tools_, _customer-service_, _marketing_, _inventory_ or _product-sourcing_;
- Overwrite the default PNG icon at [`functions/public/icon.png`](functions/public/icon.png);
- Write a good markdown description for the app at [`functions/public/description.md`](functions/public/description.md);
- Take note about [publication](#publication);

> If you are not yet a registered partner, start by filling out [this form](https://docs.google.com/forms/d/e/1FAIpQLSfd8uUsMG6N_rSFi2blGuk3Rfqi_BPp6fxschkmkdhEBVDsyw/viewform) and wait a little for our response :handshake:

## Examples

:sunglasses: [Awesome E-Com Plus applications](https://github.com/ecomplus/awesome#applications) to get inspired!

## Continuous integration

Every commit will trigger a new **deploy** (with [GitHub Actions](.github/workflows)), then your app will be accessible at:

`https://us-central1-<project-id>.cloudfunctions.net/app/` :blush:

The `assets/application.json` will be updated automatically with some package info and current Cloud Function endpoints, you can use it as body to [_Create new Application_](https://developers.e-com.plus/docs/api/#/store/applications/new-application) on Store API.

> You can skip deploy workflow by adding `[skip ci]` to the commit message.

Also, your app's access tokens to Store API will be **automatically refreshed** every 8 hours by scheduled workflow.

### Publication

If `MARKET_TOKEN` secret is set, the app will be **automatically published to [Market](https://market.e-com.plus/)** every time `assets/application.json` is changed, then merchants will be able to install it with up to date version.

## Developing and testing locally

Setup the project normally by cloning your repository and installing npm dependencies:

```bash
git clone git@github.com:$username/$app_repository.git
cd $app_repository
npm i
cd functions
npm i
```

Then you can call `npm run deploy` locally getting configuration from `.env` file on project root. You can also set a different `FIREBASE_PROJECT_ID` locally for test purpose only.

When you're done, you can publish the app by running `npm run publish:market`.

### Firebase tools

You can also use [`firebase-tools` CLI](https://firebase.google.com/docs/cli) to run tests/deploy with custom config or scripts.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [3.2.0](https://github.com/ecomplus/application-starter/compare/v3.1.2...v3.2.0) (2023-03-10)


### Features

* **import-order:** search tiny order by `numeroEcommerce` with prefix 'ecom:' ([#141](https://github.com/ecomplus/application-starter/issues/141)) ([a061685](https://github.com/ecomplus/application-starter/commit/a061685aa2f5864cfd23f7720162d9cdcdf1880c))


### Bug Fixes

* **Base Uri:** fix base Uri bug ([01f79c9](https://github.com/ecomplus/application-starter/commit/01f79c96b92bb5636f5d539b8fef2044ff0058e6))
* **clear:** properly delete old stock entries by timestamp ([2204ac2](https://github.com/ecomplus/application-starter/commit/2204ac276ab60156072ad1cd26a992bbcfdb5c39))
* **export-order:** parse buyer name to avoid invalid chars ([#127](https://github.com/ecomplus/application-starter/issues/127)) ([475fd03](https://github.com/ecomplus/application-starter/commit/475fd038060e2ff9351b86765e27f5e1c2478ac5))
* **pub-sub:** fix pubsub not repeating and authentication error ([77c91e6](https://github.com/ecomplus/application-starter/commit/77c91e65334709f2021cab3432c60b20f66a4760))
* **tiny-webhooks:** using pub/sub to queue tiny webhook events ([#130](https://github.com/ecomplus/application-starter/issues/130)) ([a6587c5](https://github.com/ecomplus/application-starter/commit/a6587c5b2b020c5295d0992354edfbdfce4799a0))

### [3.1.2](https://github.com/ecomplus/application-starter/compare/v3.1.1...v3.1.2) (2022-09-05)


### Bug Fixes

* **deps:** update all non-major dependencies ([f25957f](https://github.com/ecomplus/application-starter/commit/f25957f8a6429d4642fb72b8df7b1e2502414c97))
* **export-order:** check if has updated by tiny before send ([#120](https://github.com/ecomplus/application-starter/issues/120)) ([3f2d6b3](https://github.com/ecomplus/application-starter/commit/3f2d6b354a03458e6480e6c86c2213ac82ace7f3))
* **export-orders:** properly checking last status and prevent integration loop ([b10c49d](https://github.com/ecomplus/application-starter/commit/b10c49da66d3593128fd0bf8bef33522f54a2252))
* path function ([#122](https://github.com/ecomplus/application-starter/issues/122)) ([ccff0b4](https://github.com/ecomplus/application-starter/commit/ccff0b4e477178612697934b4038d54091a2e7ad))

### [3.1.1](https://github.com/ecomplus/application-starter/compare/v3.1.0...v3.1.1) (2022-07-08)


### Bug Fixes

* **import-product:** subtracting `saldoReservado` from quantitiy ([#113](https://github.com/ecomplus/application-starter/issues/113)) ([290f01b](https://github.com/ecomplus/application-starter/commit/290f01b5ee5285456a2e285c1a89bfe138336ba8))

## [3.1.0](https://github.com/ecomplus/application-starter/compare/v3.0.1...v3.1.0) (2022-06-28)


### Features

* **import:** import new product from tiny ([#110](https://github.com/ecomplus/application-starter/issues/110)) ([584ca51](https://github.com/ecomplus/application-starter/commit/584ca51227428b035de8052986316e0959a64281))

### [3.0.1](https://github.com/ecomplus/application-starter/compare/v3.0.0...v3.0.1) (2022-06-24)


### Bug Fixes

* **export-order:** properly handle partially paid status ([3df9692](https://github.com/ecomplus/application-starter/commit/3df96923e0f23852bd67f66422d68a2348824ffe))
* financial status already have current ([#109](https://github.com/ecomplus/application-starter/issues/109)) ([06604e7](https://github.com/ecomplus/application-starter/commit/06604e72e89abe467f594ccccf9f1ea8fc393d77))
* **parse-status:** Do not send to tiny partially_paid as paid ([3fd5508](https://github.com/ecomplus/application-starter/commit/3fd5508c8a8f2ebbe71d0866047717e490152bc6))

## [3.0.0](https://github.com/ecomplus/application-starter/compare/v2.0.0...v3.0.0) (2022-06-23)


### Bug Fixes

* **Base Uri:** fix base uri bug ([d55681b](https://github.com/ecomplus/application-starter/commit/d55681b18ff138b1942681c7c93756c1de729f48))

## [2.0.0](https://github.com/ecomplus/application-starter/compare/v1.6.5...v2.0.0) (2022-06-23)


### Features

* **parse-order-status:** add returned status as cancel ([#81](https://github.com/ecomplus/application-starter/issues/81)) ([5f8d179](https://github.com/ecomplus/application-starter/commit/5f8d179a9960a35f2cdbfdf931dfea3d43742b98))


### Bug Fixes

* **deps:** install @google-cloud/firestore again :confused: ([7e34e5b](https://github.com/ecomplus/application-starter/commit/7e34e5b678171ccf0a287a40fc3ff9981d05c8ac))
* **deps:** remove @google-cloud/firestore from direct function dependencies ([ecc73c5](https://github.com/ecomplus/application-starter/commit/ecc73c55a964a71a968b21b0390c351767113fc9))
* **deps:** update all non-major dependencies ([#82](https://github.com/ecomplus/application-starter/issues/82)) ([beb1446](https://github.com/ecomplus/application-starter/commit/beb1446a4e13cebc46c2eab8946b69fb0edf7845))
* **deps:** update all non-major dependencies ([#89](https://github.com/ecomplus/application-starter/issues/89)) ([62ea56c](https://github.com/ecomplus/application-starter/commit/62ea56c18d60223d89b58c987a4e209b6896b5d0))
* **deps:** update dependency firebase-admin to v10 ([#85](https://github.com/ecomplus/application-starter/issues/85)) ([17eee22](https://github.com/ecomplus/application-starter/commit/17eee22000a5464faaa62ea02cb6a009fe6750b3))
* **import-product:** add id on picture when parsing tiny `imagem_externa` ([#86](https://github.com/ecomplus/application-starter/issues/86)) ([d4e9932](https://github.com/ecomplus/application-starter/commit/d4e99325cf31015367bbe5aea5e7861f7593159c))
* **import-product:** query by `skus` and prefer visible/available items ([819b231](https://github.com/ecomplus/application-starter/commit/819b2315082efe3695c1094e8c8b14d3159cd807))

functions/index.js
#!/usr/bin/env node

const { functionName, operatorToken, baseUri } = require('./__env')

const path = require('path')
const recursiveReadDir = require('./lib/recursive-read-dir')

const handleEventTiny = require('./lib/pubsub/webhook-tiny')

// Firebase SDKs to setup cloud functions and access Firestore database
const admin = require('firebase-admin')
const functions = require('firebase-functions')
const functionsV2 = require('firebase-functions/v2')
admin.initializeApp()

// web server with Express
const express = require('express')
const bodyParser = require('body-parser')
const server = express()
const router = express.Router()
const routes = './routes'

// enable/disable some E-Com common routes based on configuration
const { app, procedures } = require('./ecom.config')

// handle app authentication to Store API
// https://github.com/ecomplus/application-sdk
const { ecomServerIps, setup } = require('@ecomplus/application-sdk')

server.use(bodyParser.urlencoded({ extended: false }))
server.use(bodyParser.json())

server.use((req, res, next) => {
  if (req.url.startsWith('/ecom/')) {
    // get E-Com Plus Store ID from request header
    req.storeId = parseInt(req.get('x-store-id') || req.query.store_id, 10)
    if (req.url.startsWith('/ecom/modules/')) {
      // request from Mods API
      // https://github.com/ecomclub/modules-api
      const { body } = req
      if (typeof body !== 'object' || body === null || !body.params || !body.application) {
        return res.status(406).send('Request not comming from Mods API? Invalid body')
      }
    }

    if (process.env.NODE_ENV !== 'development') {
      // check for operator token
      if (operatorToken !== req.get('x-operator-token')) {
        // last check for IP address from E-Com Plus servers
        const clientIp = req.get('x-forwarded-for') || req.connection.remoteAddress
        if (ecomServerIps.indexOf(clientIp) === -1) {
          return res.status(403).send('Who are you? Unauthorized IP address')
        }
      }
    }
  }

  // pass to the endpoint handler
  // next Express middleware
  next()
})

router.get('/', (req, res) => {
  // pretty print application body
  server.set('json spaces', 2)
  require(`${routes}/`)(req, res)
})

router.get('/_me', (req, res) => {
  server.set('json spaces', 2)
  res.send({
    host: req.get('host'),
    baseUri
  })
})

const prepareAppSdk = () => {
  // debug ecomAuth processes and ensure enable token updates by default
  process.env.ECOM_AUTH_DEBUG = 'true'
  process.env.ECOM_AUTH_UPDATE = 'enabled'
  // setup ecomAuth client with Firestore instance
  return setup(null, true, admin.firestore())
}

// base routes for E-Com Plus Store API
const routesDir = path.join(__dirname, routes)
recursiveReadDir(routesDir).filter(filepath => filepath.endsWith('.js')).forEach(filepath => {
  // set filename eg.: '/ecom/auth-callback'
  let filename = filepath.replace(routesDir, '').replace(/\.js$/i, '')
  if (path.sep !== '/') {
    filename = filename.split(path.sep).join('/')
  }
  if (filename.charAt(0) !== '/') {
    filename = `/${filename}`
  }

  // ignore some routes
  switch (filename) {
    case '/index':
      // home already set
      return
    case '/ecom/webhook':
      // don't need webhook endpoint if no procedures configured
      if (!procedures.length) {
        return
      }
      break
    default:
      if (filename.startsWith('/ecom/modules/')) {
        // check if module is enabled
        const modName = filename.split('/').pop().replace(/-/g, '_')
        if (!app.modules || !app.modules[modName] || app.modules[modName].enabled === false) {
          return
        }
      }
  }

  // expecting named exports with HTTP methods
  const methods = require(`${routes}${filename}`)
  for (const method in methods) {
    const middleware = methods[method]
    if (middleware) {
      router[method](filename, (req, res) => {
        console.log(`${method} ${filename}`)
        prepareAppSdk().then(appSdk => {
          middleware({ appSdk, admin }, req, res)
        }).catch(err => {
          console.error(err)
          res.status(500)
          res.send({
            error: 'SETUP',
            message: 'Can\'t setup `ecomAuth`, check Firebase console registers'
          })
        })
      })
    }
  }
})

server.use(router)
server.use(express.static('public'))

exports[functionName] = functions.https.onRequest(server)
exports[`${functionName}v2`] = functionsV2.https.onRequest(server)
console.log(`-- Starting '${app.title}' E-Com Plus app with Function '${functionName}'`)

// schedule update tokens job
const cron = '20 */3 * * *'
exports.updateTokens = functions.pubsub.schedule(cron).onRun(() => {
  return prepareAppSdk().then(appSdk => {
    return appSdk.updateTokens()
  })
})
console.log(`-- Sheduled update E-Com Plus tokens '${cron}'`)

// schedule sync from Tiny API to local Firestore and Store API
const syncFromTiny = require('./lib/integration/sync-from-tiny')
const syncCron = 'every 3 mins'
exports.scheduledSync = functions.pubsub.schedule(syncCron).onRun(syncFromTiny)
console.log(`-- Sheduled active sync from Tiny API '${syncCron}'`)

// delete old stored Tiny order states
const clearTinyStates = require('./lib/integration/clear-tiny-states')
const clearStatesCron = '08 09 * * *'
exports.scheduledClear = functions.pubsub.schedule(clearStatesCron).onRun(clearTinyStates)
console.log(`-- Sheduled clearing Tiny stored states '${clearStatesCron}'`)

exports.onTinyEvents = require('./lib/pubsub/create-topic')
  .createEventsFunction('tiny', handleEventTiny)
